### ISSUE TITLE
Aria Labels and HTML output

### ISSUE THREAD

**Issue by davidkartuzinski** [sbardian/gatsby-plugin-breadcrumb#58]
Hello.  Thank you for your plugin.  It works well for my uses.

However, I have two issues that are related to each other.

Currently, the plugin outputs Divs and Spans:

Like so:

```html
<div>
    <span class="breadcrumb__title">You are at: </span>
            <div class="breadcrumb">
                   <a class="breadcrumb__link" href="/">Home</a>
                   <span class="breadcrumb__separator"></span>
             </div>
      <div class="breadcrumb">
             <span class="breadcrumb__link__disabled">blog</span>
             <span class="breadcrumb__separator"></span>
       </div>
</div>
```

Issue #1:

Per https://www.w3.org/TR/wai-aria-practices-1.1/examples/breadcrumb/index.html. 

A more accessible pattern is:

```html
<nav aria-label="Breadcrumb" class="breadcrumb">
  <ol>
    <li>
      <a href="../../">
        WAI-ARIA Authoring Practices 1.1
      </a>
    </li>
    <li>
      <a href="../../#aria_ex">
        Design Patterns
      </a>
    </li>
    <li>
      <a href="../../#breadcrumb">
        Breadcrumb Pattern
      </a>
    </li>
    <li>
      <a href="./index.html" aria-current="page">
        Breadcrumb Example
      </a>
    </li>
  </ol>
</nav>

```

Like this example, I was able to easily wrap the Breadcrumbs in a Nav element. In doing that I can add the "aria-label='breadcrumb'" to the nav element.

**Question: 1.** The first issue is how can I go about adding "aria-current="page" to the last link? But I suppose only if it is NOT disabled?

**Question: 2.** How could I go about changing the HTML structure? Is the only way to modify the https://github.com/sbardian/gatsby-plugin-breadcrumb/blob/develop/src/components/auto-gen-crumb.js page? 

And lastly, would you be interested in a Pull Request for changing the HTML to follow these guidelines? I don't know any reason why anyone would prefer divs? An Ordered List, if you think about it, makes a lot more sense. 

Again thanks, your plugin is really good. And the other options you've provided just work!

**Comment by sbardian** [comment]
@davidkartuzinski Thanks for the issue!

#1 - I think if you pass aria-current prop it will be spread across the div with the `breadcrumb` class.  Pretty sure I spread any extra props there.

#2 - Currently the only way would be to not use the `<Breadcrumb>` component and to write your own, passing the data generated by the plugin using autogen to your own "breadcrumb" component using nav/ol/li HTML. 

Let me look into updating the div/span setup to nav/ol/li.  This would be a nice update and I would like to do it for both click tracking and autogen.  If you need this before I can push an update feel free to write your own component for the breadcrumb data, if not check back here and I'll update with my findings.  Thanks again.

**Comment by davidkartuzinski** [comment]
> #1 - I think if you pass aria-current prop it will be spread across the div with the `breadcrumb` class. Pretty sure I spread any extra props there.

@sbardian - Thank you.  Generally speaking I can wait. I am still working on making my own theme, and it will be better for me to have people use your plugin (which is great), than something I try and throw together.

As to #1 - I am not sure I explained it well. 

If the breadcrumb is:

*Home > Blog > **Awesome-Article***

The generated HTML, would have  `aria-current="page"` on the last `a` tag | which would be the current page. In this example, **Awesome-Article**. 

---

If I did explain it correctly, then are you saying to pass the prop from`gatsby-config.js` file or through the component?

Thanks for your quick response.

**Comment by sbardian** [comment]
@davidkartuzinski hmm looking at the code, I think passing it as a prop (to `<Breadcrumb>` component) would actually spread / add it to all the "crumbs" now that I'm looking at it.  So probably wouldn't work.  I'm working on updated to nav/ol/li now though, we will see how it goes.

**Comment by davidkartuzinski** [comment]
Thank you very much.  I appreciate it. Please let me know if I can help with anything?

**Comment by sbardian** [comment]
@davidkartuzinski 
I think I have a release ready for this update, but wanted to get your thoughts on some of the updates.  Currently the HTML will come out looking something like this: 

```
<span class="breadcrumb__title"></span>
<nav class="breadcrumb" aria-label="Breadcrumb" style="display: inline;">
  <ol
    class="breadcrumb__list"
    style="display: block; list-style: outside none none;"
  >
    <li class="breadcrumb__list__item" style="display: inline;">
      <a class="breadcrumb__link" href="/">Home</a>
    </li>
    <span
      class="breadcrumb__separator"
      style="display: inline-block; font-size: 16pt; margin: 0px 0.25em;"
      >/</span
    >
    <li class="breadcrumb__list__item" style="display: inline;">
      <span class="breadcrumb__link__disabled">Books</span>
    </li>
    <span
      class="breadcrumb__separator"
      style="display: inline-block; font-size: 16pt; margin: 0px 0.25em;"
      >/</span
    >
    <li class="breadcrumb__list__item" style="display: inline;">
      <a
        aria-current="page"
        class="breadcrumb__link breadcrumb__link__active"
        href="/book/919c102c-3981-5778-b335-e70cd60001f5"
        >The Wheel of Time</a
      >
    </li>
  </ol>
</nav>
```

I think this matches up pretty well, keeps class names where they should be, etc.  

The only issue I had was the Breadcrumb Separator (the `<span>` after each `<a>`/ link).  Instead of including a CSS-in-JS library I use the React style prop, which does not support pseudo elements (ie: ::before).  Reading around in React they say just "add an element/span" before or after / where you need it.  So that is what I have done here.  

Reading the W3C WAI-ARIA docs, it sounds like this might just cause some "noise"?  Not sure if this is acceptable or not? 

From the W3C docs: 
> To prevent screen reader announcement of the visual separators between links, 
> they are added via CSS:

> The separators are part of the visual presentation that signifies the breadcrumb 
> trail, which is already semantically represented by the nav element with its label of
> "Breadcrumb". So, using a display technique that is not represented in the accessibility 
> tree used by screen readers prevents redundant and potentially distracting verbosity. 

I'm not great in the accessibility arena (hence this issue) so I thought I would ask your input.  The only other options I can think of are a much larger update to pull out inline CSS-in-JS and publishing a CSS file along side the library for consumers to use/not use, or striping out inline CSS-in-JS and leaving it up to the consumer to style the breadcrumb based on the provided class names (basically the useClassNames plugin option).  Let me know what you think.

**Comment by davidkartuzinski** [comment]
@sbardian - Hi, and thank you for this amazing work!

*About the Inline CSS*

I understand your question about the CSS.  In my opinion, I think it's best to strip out the inline CSS.  A lot of people like Styled Components but also Emotion.  Some like a global.css file while many also like modular CSS, and still others like libraries like Semantic UI, Bootstrap, Tailwinds, etc.  

I say remove the CSS from inline. That allows anyone to use whatever method they want to style it. 

I would keep the classes in there though.  These could be helpful. 

*About the Span issue*

But that leaves us the problem of how to add a separator.  The span can stay but like this: 

https://www.aditus.io/patterns/breadcrumbs/#example-2-separators-in-html

Here they say to take them out, but the above link just makes it clear to add `<span aria-hidden="true">→<span>`

https://www.aditus.io/patterns/breadcrumbs/#4-optional-make-separators-decorative

*About how to add CSS*

Let's give a CSS code snippet, that people can copy/paste.  For example, like https://github.com/negomi/react-burger-menu#styling - I use this library and I just copy/paste it into my component (if using styled-components) or my global CSS file or whatever. 

We need:

```CSS
.breadcrumb ol {
    display: flex;
    flex-direction: column; // mobile first make the <li> into a column
}
@media screen and (min-width: 500px) {
       .breadcrumb ol { 
        flex-direction: row; // after 500px and larger make the breadcrumb into a row
    }
}
```

I could contribue some CSS snippets that people can cut/paste and use however they want.

*About testing the final output*

We could again test the output of the breadcrumbs against - https://achecker.ca/checker/index.php - I pasted your HTML above in here, and it already passed (basically).  

I would like to put up a demo site on netlify, that we can further test with this:
https://wave.webaim.org/sitewide

The other side of the coin, is the screenreader.  That is the **real** test!  I can test it on my mac and linux machine. I don't have windows. Here are the tools. https://soap.stanford.edu/tips/screen-reader-testing

I'd be willing to get onto a Windows machine and test it though. 

So, basically, I am proposing:

1. Strip out the inline CSS
2. add the proper span code like so, `<span aria-hidden="true">→<span>`
3. I can propose CSS snippets and something to add to the docs.
4. Put up a demo site with the breadcrumbs and test it and leave it up so people can see for themselves. 
5. Update docs.

I am not trying to make this a bigger deal that it needs to be, but I think we are really close to have this aspect solved well!

Please let me know what you think?  If you like the ideas, please make a branch that I can fork and get onto.  

:)

-DK

**Comment by sbardian** [comment]
@davidkartuzinski  I agree inline CSS should probably go. And publishing CSS along side the plugin for consumers to use sounds like a good option to me.  

I have added the `aria-hidden="true"` attribute to the breadcrumb separator spans, which should handle the 'noise' issue, and I think this gets this update to where I can publish it, thoughts? 

Then I can start work on the larger update of pulling inline CSS and publishing a CSS file along side the plugin.  This can/should include your 1-5 steps, but I will also have to account for other things and plugin options/component props that are currently available (useClassNames/crumbWrapperStyle/crumbStyle/crumbActiveStyle/etc.).

This sound acceptable?  I would like to get out a more accessible version following the best practices you present/point out, and then work on the CSS issues after as it might take some time for me to get that all sorted.

**Comment by davidkartuzinski** [comment]
@sbardian Yes, please! The CSS is less of a big deal. 

Please tell me how I can help?

**Comment by sbardian** [comment]
@davidkartuzinski thanks! I'll get it published.  Let me get the core parts of the updates to the plugin going, then I will take the first shot at a CSS update and we can work through specifics on that together.

**Comment by sbardian** [comment]
v8.0.0 has been cut.  Give it a try and make sure I didn't screw up anywhere :) hopefully all is well.  I'll get to work on the larger update this week and then get stuff pushed to dev where we can work through details together.  I'll leave this issue open for now while we work through that larger update. 
thanks again

**Comment by sbardian** [comment]
@davidkartuzinski I have pushed a new branch (styles: https://github.com/sbardian/gatsby-plugin-breadcrumb/tree/styles) with the start to the work for removing inline styles.  Most of what we are going to care about is in:

* src/components/auto-gen-crumb.js
* src/components/click-tracking-crumb.js
* /gatsby-plugin-breadcrumb.css
* /README.md

I think the components should be fine, I pulled out all props dealing with styles and any default style props applied to elements.  We still have the nav/ol/li HTML structure/etc.  I left the `<span>` for the separator since we have the `aria-ignore="true"` attribute applied.  Leaving this still allows users to pass their own custom separator if they wish.  

I threw in some basic CSS in `gatsby-plugin-breadcrumb.css` so it looks right at least on my site. . . May want to pull out some of the text options, sizes, etc., but I wanted to get a starting point to match sorta what I had inline.

I also updated the README.md to remove things I pulled and also to start talking about _how_ users can style their breadcrumbs.  

Give it a look when you have some time and let me know what you think.  I may plug a little more but will update here if anything changes drastically.

**Comment by davidkartuzinski** [comment]
@sbardian Hello! Sorry it took all day to get back to you.  I installed the new version and it's great! 
I can't tell if you intended this, but I am not getting any CSS coming through. I can't find the /gatsby-plugin-breadcrumb.css file in the node_modules. Or with search.

Flex and Flex wrap, list style none <---- should be really all that is needed as a base minimum. If that is the way you want to go. Which I like as I can style anyway I want without too much trouble. 

So in case it wasn't clear, I think this is really great, and now I can publish my core-theme with this as a dependency which makes me soo happy!

Thank you, and please let me know if you want to me to do anything else.

-DK

ps. You should get yourself some exposure and write a little article on how you brought your plugin in step with the Gatsby Teams Core Values.  They'll publish it on their blog!

**Comment by sbardian** [comment]
@davidkartuzinski  the release v8.0.0 does not have those changes.  That will be part of a larger release removing inline styles, that I'm in the middle of.  See the new branch called "styles" here https://github.com/sbardian/gatsby-plugin-breadcrumb/tree/styles

I will continue to tweak and get this ready for a release (v9.0.0).  I think hitting all the steps you mentioned (1-5 above) would be nice.  I think I'm close.  Watch for that release in the future!  Feel free to throw in any comments / etc. on that branch and I will give them a look. 

haha Thanks for the offer on the blog, I'll look into it!

**Comment by sbardian** [comment]
@davidkartuzinski reading your comment again. . . if you are not getting _any_ css on the breadcrumb using v8.0.0 then something might be wrong.  That should have just updated the HTML to use nav/ol/li, keeping inline styles and style props. . . My project is still getting the inline props but in the new HTML structure. . .

**Comment by davidkartuzinski** [comment]
@sbardian - You're fast! 

Yeah, I am not getting any CSS. I am getting CSS Class though. And everything is working.
![Screenshot 2020-02-24 at 21 46 49](https://user-images.githubusercontent.com/29503118/75190056-d1ac8a00-574f-11ea-86d6-8674b946aed7.png)


```html
<nav>
    <nav aria-label="Breadcrumb">
        <span class="breadcrumb__title">You are here: </span>
         <nav class="breadcrumb" aria-label="Breadcrumb">
              <ol class="breadcrumb__list">
                    <li class="breadcrumb__list__item">
                         <a class="breadcrumb__link" href="/">Home</a>
                     </li>
                      <span class="breadcrumb__separator" aria-hidden="true"> &gt; </span>
                       <li class="breadcrumb__list__item">
                           <a aria-current="page" class="breadcrumb__link breadcrumb__link__active" href="/tags">tags</a>
                        </li>
                   </ol>
             </nav>
     </nav>
</nav>
```

If you add too much more CSS it could get cumbersome for anyone (like me) who is including your plugin as a starter or theme.  BUT for accessibility purposes, I would definitely include 
`:focus-within, &:hover` from our friend https://www.scottohara.me/blog/2017/05/14/focus-within.html

A real life example, is in the Gatsby Docs here: https://www.gatsbyjs.org/tutorial/building-a-theme/#make-themes-extendable-with-gatsby-plugin-theme-ui --- You just need to scroll down a little bit.

===

Otherwise, if you want I can ghost write your blog post for you. And then you can edit it and make it your own.  I did this sort of thing a lot when I was Developer Advocate at Strapi.io.

Again thank you! And I look forward to your updates.

_DK

**Comment by sbardian** [comment]
@davidkartuzinski  ahhh you are using `useClassNames`.  That disables any inline CSS and default styling, so we are all good.  That is how it should be. 

I will look into the accessibility includes you mention, thanks! 

Wow, yeah if you want to ghost write it that would be awesome!  Let me know. 

Thanks!

**Comment by davidkartuzinski** [comment]
@sbardian Okay okay. So when I set it to False, your CSS comes through. I can confirm.

But not to change the subject, but is this code below supposed to do anything? I forgot I did this and it's not working for me?

```
        crumbLabelUpdates: [
          {
            crumbStyle: { color: "blue" },
            crumbActiveStyle: { color: "red" }
          }
        ],
```

I think when you do your final release I can get the article to you.  Otherwise, you'll get too much attention haha

When you submit a PR with them you get automatic swag, which I love! You will need to use your GitHub account though. It's super easy.

-DK

**Comment by sbardian** [comment]
@davidkartuzinski so that is for updating parts of a path that are not page endpoints.  So say you have the following breadcrumb: 

Home / books / Wheel of Time
There are ways of updating the "Home" value (ie: plugin option autoGenHomeLabel: `Root`) and you can update the end (Wheel of Time) via the crumbLabel prop  to the `<Breadcrumb>` component.  To be able to catch the inner section of the breadcrumb, that may or may not have an actual page you use the following plugin option: 

```
        // crumbLabelUpdates: optional, update specific crumbLabels in the path
        crumbLabelUpdates: [
          {
            pathname: '/book',
            crumbLabel: 'Books'
          }
        ]
```

I can report back when I release v9.0.0 pulling out the inline css styles.  I have contributed to gatsby in the past, and have collected a little swag myself :) 

Are you thinking I should use the Gatsby Theme UI stuff for people to override the default css file I release in v9.0.0?  Instead of just having them import their own css targeting the class names provided?

**Comment by davidkartuzinski** [comment]
> so that is for updating parts of a path that are not page endpoints.

Okay. So I am not using them correctly! Makes sense haha. Thank you for the explanation.

I am happy that you got yourself some swag! So you know then how easy and fun it is. I will look out for sure for your update.  And then send you the article.

I might do a version of the theme using Gatsby Theme UI - I am not so much of an emotion css guy, and my understanding is that Gatsby Theme UI is basically powered by emotion.  I call my theme "Naked" as it is to function as a base to other themes, and then have different CSS based versions that is separate from the core and then even additional CMS based versions. 

All this stuff with accessibility, SEO, core widgets, schema, etc.  Should just come out of the box when we install Gatsby.  There are Best Practices and I think the next phase for Gatsby is to make it harder to not do good SEO, Schema etc... than just to do it properly out of the gate.  Unfortunately, schema and accessibility are hard. I think Ghost is the only theme that is doing Schema correctly.   

Like with your plugin... which now makes this vital part of the UX accessible and with the added bonus we get SEO bonus.  

Thanks again for your hard work. 

-DK

**Comment by sbardian** [comment]
@davidkartuzinski sounds good!  I am going to close this issue and create a new one for striping out inline css/providing a default css file/etc. (new issue https://github.com/sbardian/gatsby-plugin-breadcrumb/issues/59).  Thanks for all your help, feel free to follow there to see updates or just keep an eye out for v9. Enjoy!

**Comment by sbardian** [comment]
@davidkartuzinski was looking closer at your component, you might be getting double `<nav>` elements, as the breadcrumb component already wraps it for you, and it looks like you are wrapping it as well.  Just something I noticed you might check on. 

```
const NakedBreadcrumb = ({ crumblabel, crumbs }) => {
  return (
    <nav aria-label="Breadcrumb">
      <Breadcrumb
        crumbs={crumbs}
        title="You are here: "
        crumbSeparator=" > "
        crumbLabel={crumblabel}
      />
    </nav>
  );
};
```

**Comment by davidkartuzinski** [comment]
@sbardian  - Yes. I haven't pushed my local up to GitHub because I borked it and I need to fix it first. But yes I made the updates.  I will push the updates later and let you know.  But I had already remove the Nav around the breadcrumbs. Thank you very much!
